Another attempt at multi-layer item rendering.

gigaherz opened the pull request:
This is the third internal redesign, but the first one I feel comfortable showing.

## Intended use cases

1. Drawing multi-layer BlockItems in an equivalent way to the placed blocks.
2. Improving the fullbright item support by allowing fullbright quads to be drawn using an unlit render type.
3. Improving custom model support (eg. dynamic bucket models) by allowing models to be drawn without transparency sorting.

## What these use cases require

1. The ability to draw quads in multiple render types. (multi-layer blocks, custom models)
2. The ability to repeat render types in a well-defined order. (layered items)

## The solution

By returning a `List<Pair<IBakedModel, RenderType>>`, the code allows both use cases of the same model being drawn with multiple render types (needed by the existing multi-layer block system), and other models being drawn with a specific render type (needed by layered items with mixed fullbright quads).

gigaherz committed:
Implement multi-layer item rendering.

pupnewfster commented:
Can this system be used to fix the bucket model loader having the fluid make the bucket transparent?

gigaherz committed:
Cleanup ItemMultiLayerBakedModel construction by introducing a Builder.
Make use of ItemMultiLayerBakedModel from the DynamicBucketModel, to allow custom render type usage, including illuminating fluids drawing as fullbright.

gigaherz committed:
Fix bucket flip being using the wrong coordinate origin.

gigaherz commented:
Dynamic bucket model is now capable of showing both transparent fluids:
![transparent fluid](https://cdn.discordapp.com/attachments/679070190721433659/719182825605300254/unknown.png)

And illuminated fluids:
![illuminated fluid](https://cdn.discordapp.com/attachments/679070190721433659/719122779290730536/unknown.png)

And multi-layer block models will work as intended when used in the corresponding BlockItem:
![multi-layer](https://user-images.githubusercontent.com/685851/83970718-79049580-a8d7-11ea-86d5-8663a242de31.png)
(left: vanilla slime block, right: custom test model)

gigaherz committed:
Make fluid luminosity detection opt-out.

cech12 commented:
Is there a plan to merge it? :)

gigaherz commented:
A derivative of this was merged into 1.16 a while ago. In order to merge this, we would need to make sure it lines up with the current implementation in 1.16, while also making sure it doesn't introduce any breaking change into the LTS version.
I don't currently have the energy/interest to do that backport.

Edited to add: If someone wants to start from the 1.16 code with this as a reference, and see if they can make a proper backport, feel free!

gigaherz commented:
Oops I forgot about this. I don't really have any interest in backporting, so no sense keeping this open.
